src/main/java/cam72cam/immersiverailroading/model/ModelState.java

```diff
@@ -225,25 +226,87 @@ private boolean hasGroupFlag(String group, String filter) {
         return false;
     }
 
+    private static class Opacity {
+        final List<String> opaque = new ArrayList<>();
+        final List<String> transparent = new ArrayList<>();
+
+        void add(boolean isTransparent, String group) {
+            if (isTransparent) {
+                transparent.add(group);
+            } else {
+                opaque.add(group);
+            }
+        }
+    }
+
+    private static class RenderGroup {
+        public Opacity animated = new Opacity();
+        public Opacity notAnimated = new Opacity();
+        private final Set<String> animatedGroups;
+
+        public RenderGroup(Map<String, Matrix4> animatedGroups) {
+            this.animatedGroups = animatedGroups.keySet();
+        }
+
+        public void add(String group) {
+            boolean hasAlpha = group.contains("ALPHA");
+
+            if (animatedGroups.contains(group)) {
+                animated.add(hasAlpha, group);
+            } else {
+                notAnimated.add(hasAlpha, group);
+            }
+        }
+    }
+
+    private static class LightKey {
+        final int sky;
+        final int block;
+        LightKey(float sky, float block) {
+            this.sky = Float.floatToRawIntBits(sky);
+            this.block = Float.floatToRawIntBits(block);
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * sky + block;
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (this == obj) return true;
+            if (!(obj instanceof LightKey)) return false;
+            LightKey k = (LightKey) obj;
+            return sky == k.sky && block == k.block;
+        }
+    }
+
+    private static final BlendMode ALPHA_BLEND = new BlendMode(BlendMode.GL_SRC_ALPHA, BlendMode.GL_ONE_MINUS_SRC_ALPHA);
+    private final List<String> singletonList = new ArrayList<>();
+    private final Map<LightKey, RenderGroup> levels = new Object2ObjectOpenHashMap<>();
```

Comment:
This does not need to be cached, the existing render pipeline creates a new ModelState frequently (for better or worse)